### PR TITLE
Project history

### PR DESCRIPTION
**Pull request by khoroshevskyi**
Changes:
- Added functionality that saves project history after save button was clicked

TODO:
- [x] Add tests
- [x] Disable history for huge projects

### CODE REVIEW

File: pepdbagent/db_utils.py [hunk 1]
```diff
@@ -119,6 +121,10 @@ class Projects(Base):
 
     namespace_mapping: Mapped["User"] = relationship("User", back_populates="projects_mapping")
 
+    history_mapping: Mapped[List["HistoryProjects"]] = relationship(
+        back_populates="project_mapping", cascade="all, delete-orphan"
+    )  # TODO: check if cascade is correct
```
Comment (nleroy917) [inline pepdbagent/db_utils.py:126]: Does this need implementing in this PR? I see its a TODO

File: pepdbagent/modules/project.py [hunk 2]
```diff
@@ -558,13 +605,27 @@ def update(
                             f"pephub_sample_id '{PEPHUB_SAMPLE_ID_KEY}' is missing in samples."
                             f"Please provide it to update samples, or use overwrite method."
                         )
+                    if len(update_dict["samples"]) > MAX_HISTORY_SAMPLES_NUMBER:
+                        _LOGGER.warning(
+                            f"Number of samples in the project exceeds the limit of {MAX_HISTORY_SAMPLES_NUMBER}."
+                            f"Samples won't be updated."
+                        )
+                        new_history = None
```
Comment (nleroy917) [inline pepdbagent/modules/project.py:613]: If it's not too complicated, can we just delete the oldest one and then insert?

Comment (nleroy917) [reply pepdbagent/modules/project.py:613]: Also, does this warrant an `Exception`? I guess a warning is fine, but if the user is not paying attention they might not realize their samples histories aren't being updated 😅

File: pepdbagent/modules/project.py [hunk 3]
```diff
@@ -970,3 +1076,235 @@ def get_samples(
...
+                    pass
+
+        if raw:
+            return {
+                CONFIG_KEY: project_config or project_mapping.config,
+                SAMPLE_RAW_DICT_KEY: ordered_samples_list,
+                SUBSAMPLE_RAW_LIST_KEY: self.get_subsamples(namespace, name, tag),
+            }
+        return peppy.Project.from_dict(
+            pep_dictionary={
```
Comment (nleroy917) [inline pepdbagent/modules/project.py:1224]: linting? `pep_dictionary = {` ?? Not sure

File: pepdbagent/modules/project.py [hunk 4]
```diff
@@ -970,3 +1076,235 @@ def get_samples(
...
+                CONFIG_KEY: project_config or project_mapping.config,
+                SAMPLE_RAW_DICT_KEY: ordered_samples_list,
+                SUBSAMPLE_RAW_LIST_KEY: self.get_subsamples(namespace, name, tag),
+            }
+        )
+
+    @staticmethod
+    def _apply_history_changes(sample_list: dict, change: HistoryProjects) -> dict:
```
Comment (nleroy917) [inline pepdbagent/modules/project.py:1232]: Shoulld `sample_list: dict` be a `list[dict]` ?

Comment (nleroy917) [reply pepdbagent/modules/project.py:1232]: I'm not sure, just checking

**Review by nleroy917 — APPROVED** [review]
Yeah this looks great. Just small things. The only real one is my suggestion for deleting a row before inserting history after warning that they've hit their maximum

File: pepdbagent/db_utils.py (hunk 1 above)
Comment (khoroshevskyi) [reply pepdbagent/db_utils.py:126]: done

File: pepdbagent/modules/project.py (hunk 4 above)
Comment (khoroshevskyi) [reply pepdbagent/modules/project.py:1238]: Fixed

**Pull request merged**